Better editing of the standard library in coqtop/PG

JasonGross opened the pull request:
Change a proof that only compiled non-interactively in `Coq.Init.Logic`, and add `.dir-locals.el` files to set up PG to allow easy editing of the standard library.

@cpitclaudel might have useful comments about the `.dir-locals.el` setup.

silene commented:
No way! If there is a bug, we better fix it rather than modifying the proof to hide it.

JasonGross commented:
The "bug" is that Coq requires `Coq.Init.Logic.eq` to exist before using `replace foo with bar`.  I suspect we could instead add the following to the top of this file:
```coq
(* -*- coq-prog-args: ("-emacs" "-noinit" "-R" ".." "Coq" "-topname" "Logic") -*- *)
```

Alternatively, we could modify ProofGeneral to pass the current file name as `-topname` when it's a valid module name (which PG would have to determine for itself) (@cpitclaudel thoughts on this?).  Would one of these be preferable?

herbelin commented:
Yes, the standard trick is to use "-top Coq.Init.Logic" so that references to Logic made by tactics work.

At some time, there were a similar issue with `tauto` but this was also found to be a source of confusion by some people, so 5651885f6 eventually got rid of it.

I personally have no strong opinion on the issue. A model satisfactory to the mind could be to have an include of the code of "replace" done within Logic.v rather than in Notations.v, but somehow, it may be as simple to prosaically remove the call to replace.

cpitclaudel commented on theories/.dir-locals.el:
```diff
@@ -0,0 +1,6 @@
+						buffer-file-name ".dir-locals.el")))
+			(make-local-variable 'coq-prog-args)
+			(make-local-variable 'coq-prog-name)
+			(setq coq-prog-args `("-emacs" "-coqlib" ,(expand-file-name "..") "-R" ,(expand-file-name ".") "Coq"))
```
Just use setq-local instead of make-local-variable + setq

cpitclaudel replied:
Also, do you need -emacs in the coq-prog-args?

psteckler replied:
The coq-prog-name assumes a particular directory structure, which can be overridden when running `configure`.

JasonGross replied:
> use setq-local

Done

> do you need -emacs

Apparently not; removed

JasonGross replied:
This is true.  However, the system-wide Coq will almost certainly not work; I got interface mismatches on things like `nat_syntax_plugin`.  So users that want to edit the std lib currently have to manually set the coq binary.  This change fixes that problem for the common case of the default directory structure, and I'm not sure it's worth the overhead to have `./configure` autogenerate `.dir-locals.el`.  What do others think?

psteckler replied:
It's a few lines of code to add to `configure.ml`, not much overhead, I think.

cpitclaudel commented on theories/Init/.dir-locals.el:
```diff
@@ -0,0 +1,4 @@
+((coq-mode . ((eval . (let ((default-directory (locate-dominating-file
+						buffer-file-name ".dir-locals.el")))
+			(make-local-variable 'coq-prog-args)
+			(setq coq-prog-args `("-emacs" "-R" ,(expand-file-name "..") "Coq" "-noinit")))))))
```
Why don't we add a CoqProject instead of this?

JasonGross replied:
Because `-noinit` is required for files in the `Init` directory, and wrong for every other file.

cpitclaudel replied:
Sorry, I should have posted this comment under the other .dir-locals.el.

gares commented:
Maybe a better fix is to put `eq` into another file that is compiled before the one that needs the `replace` tactic.
So, all in all, the fix by Jason is not that bad ;-)

JasonGross commented:
@cpitclaudel said in an outdated diff:
> Why don't we add a CoqProject instead of this?

`_CoqProject` won't let us set the coq binary, IIRC, and also traditionally has all of the .v files listed; using .dir-locals.el puts less maintenance burden on the devs, I believe, though I suppose I could just do
```
-R . Coq
-coqlib ..
```
or something in the `_CoqProject`.  Would that be preferable?


@gares moving `eq` seems iffy, from a compatibility perspective.  Certainly, we'd at least need notations so that `Logic.eq` still resolves.  I wanted to make the minimally invasive change, here.

Zimmi48 commented:
As remarked by Christian Doczkal just now on the Coq-Club mailing list, one can add a `_CoqProject` that just lists the option `-R . Coq` to be passed to Proof-General and CoqIDE, without listing the `.v` files. Of course it won't be used to generate the Makefile but it has the advantage to be useful both for Proof-General and CoqIDE and it does not prevent to add more stuff in `theories/.dir.local.el`, does it?

JasonGross commented:
@Zimmi48 That does not work.  The `_CoqProject` file will override a `.dir-locals.el` in the same directory, and there is no way to get PG to pick up a valid `-coqlib` from `_CoqProject` (because it'll take the path relative to the current directory, not relative to the directory of `_CoqProject`).

I've added `theories/Init/_CoqProject` and removed `theories/Init/.dir-locals.el`, though, since that works.

JasonGross commented:
I've rebased this on trunk.  What's the status of merging this?

Zimmi48 commented:
LGTM although the commits need some squashing.

Would there also be a way to edit interactively `theories/Classes/Morphisms.v` and `theories/Classes/CMorphisms.v` without passing manually the options `-top Coq.Classes.Morphisms` / `-top Coq.Classes.CMorphisms` to coqtop / CoqIDE?

JasonGross commented:
Should I squash, or will the devs squash on merging it?

> Would there also be a way to edit interactively 

I'll take a look, but I think that's an issue for a separate PR, since it'll probably happen by editing those files.

maximedenes commented:
@JasonGross Can you squash? I'll merge once it is done. Thanks!

JasonGross commented:
@maximedenes Rebased and squashed

JasonGross commented:
@Zimmi48 
> Would there also be a way to edit interactively theories/Classes/Morphisms.v and theories/Classes/CMorphisms.v without passing manually the options -top Coq.Classes.Morphisms / -top Coq.Classes.CMorphisms to coqtop / CoqIDE?

I don't think it's that easy.  Most of the uses of `rewrite` can be replaced by `apply`, but the `rewrite Hxz` on the last line of `PartialOrder_StrictOrder` is hard to do manually.  However, I can update the line at the top so that emacs picks up, e.g., `-top Coq.Classes.Morphisms` automatically; will that satisfy you?

Zimmi48 commented:
@JasonGross Oh yes, I was not actually thinking that we should remove the uses of `rewrite` there. So I'm satisfied with the solution you propose. It will solve the problem only in Emacs, but it's also easier to pass this option to CoqIDE, and the line will give the hint as to what option is needed.

Zimmi48 commented:
I still have problems when trying to edit the `.v` files in plugins with Emacs, if ever you have the time to look at it.

JasonGross commented:
What are the proper flags to pass to the plugins files?  (What does the Makefile pass to build files in that directory?)

Zimmi48 commented:
@JasonGross Not sure but from the error I see in Emacs, files in say `plugins/foo` are compiled under the prefix `Coq.foo` (similarly to those in `theories/foo`).
